src/hooks/useOpenAI.tsx

```diff
@@ -48,63 +52,142 @@ export interface OpenAIStreamingProps {
+    content,
+    role,
+  }: ChatMessage): OpenAIChatMessage => ({ content, role })
+
+  const createChatMessage = ({
+    content,
+    role,
+    ...restOfParams
+  }: ChatMessageParams): ChatMessage => ({
+    content,
+    role,
+    timestamp: restOfParams.timestamp || Date.now(),
+    meta: {
+      loading: false,
+      responseTime: '',
+      chunks: [],
+      ...restOfParams.meta,
+    },
+  })
+
+  // Initial value of message
   const systemMessage = createChatMessage({
     content: systemPrompt || '',
     role: ChatRole.SYSTEM,
   })
 
-  const [storedMessages, setStoredMessages] = useStorage<ChatMessage[]>(
-    'CHAT_MESSAGES',
-    [systemMessage],
+  // getting required data
+  const uuid = getNewUUID()
+  const initialChatHistoryValue = [
+    {
+      id: uuid,
+      url: window.location.href,
+      timestamp: Date.now(),
+      ChatMessages: [],
+    },
+  ]
+  const [chatHistory, setChatHistory] = useStorage<ChatHistory[]>(
+    'CHAT_HISTORY',
+    initialChatHistoryValue,
     'local',
   )
-  const [messages, setMessages] = useState<ChatMessage[]>([systemMessage])
+  const [currentUrl, setCurrentUrl] = useState<string>('')
+
+  /**
+   * @param content
+   * @param role
+   * @param ...restOfParams
+   * @return {*}  {ChatMessage}
+   */
+
+  // const initialChatMessage = currentMessage.length > 0 ? currentMessage : [systemMessage]
+
   const [loading, setLoading] = useState<boolean>(false)
+  const [messages, setMessages] = useState<ChatMessage[]>([systemMessage])
 
   useEffect(() => {
-    if (storedMessages.length > 1 && messages.length <= 1) {
-      setMessages(storedMessages)
+    const setState = async () => {
+      const url = await getActiveTabUrl();
+      console.log(url);
+      setCurrentUrl(url)
     }
-  }, [storedMessages])
+    setState()
+  }, [messages])
 
   useEffect(() => {
+
+
     if (messages.length > 1 && !messages[messages.length - 1].meta.loading) {
-      setStoredMessages(messages)
-      setLoading(false)
+      setChatHistory((prev) => {
+        let index = prev.findIndex(
+          (chatMessage) => chatMessage.url === currentUrl,
+        )
+        let update = prev
+        if (index === -1) {
+          index = update.length
+          console.log(messages.length, "if block");
+          update[index] = {
+            id: getNewUUID(),
+            url: currentUrl,
+            timestamp: messages[0].timestamp,
+            ChatMessages: messages,
+          }
+        } else {
+          if (update[index].timestamp !== messages[messages.length - 2].timestamp && messages.length >= 3) {
+            update[index] = {
+              ...update[index],
+              timestamp: messages[messages.length - 2].timestamp,
+              ChatMessages: [
+                ...update[index].ChatMessages,
+                messages[messages.length - 2],
+                messages[messages.length - 1],
+              ],
+            }
+          }
+        }
+        return update
+      })
```

Comment:
wow, probably we should extract this logic to its own function

Reply:
we can extract this logic into a function called createOrUpdateChatHistory